model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java

```diff
@@ -790,6 +794,48 @@ public void removeClientScopes(RealmModel realm) {
         realm.getClientScopesStream().map(ClientScopeModel::getId).forEach(id -> this.removeClientScope(realm, id));
     }
 
+    @Override
+    public void addClientScopes(ClientModel client, Set<ClientScopeModel> clientScopes, boolean defaultScope) {
+        Map<String, ClientScopeModel> existingClientScopes = getClientScopes(client, defaultScope, false);
+
+        clientScopes.stream()
+            .filter(clientScope -> ! existingClientScopes.containsKey(clientScope.getName()))
+            .forEach(clientScope -> {
+                ClientScopeClientMappingEntity entity = new ClientScopeClientMappingEntity();
+                entity.setClientScopeId(clientScope.getId());
+                entity.setClientId(client.getId());
+                entity.setDefaultScope(defaultScope);
+                em.persist(entity);
+                em.flush();
```

Comment:
Can the `flush` be moved to after the stream processing (after line 811)?

Reply:
I haven't found a way to avoid having `flush` inside `forEach`. When I move the `flush` out, I got 

```
java.lang.RuntimeException: RESTEASY003325: Failed to construct public org.keycloak.services.resources.KeycloakApplication()
...
Caused by: java.lang.RuntimeException: org.keycloak.models.ModelException: org.hibernate.AssertionFailure: possible non-threadsafe access to session
...
```